File: 2021W04/violetxf/pgr_49191.cpp
/*
    순위
    https://programmers.co.kr/learn/courses/30/lessons/49191
*/
#include <string>
#include <vector>
using namespace std;

int solution(int n, vector<vector<int>> results) {
    int answer = 0;
    vector<vector<bool>> matchData(n+1, vector<bool>(n+1, false));
    for(auto k : results){
        int w = k[0];
        int l = k[1];
        matchData[w][l] = true;
    }
    for(int i=1; i<=n; i++){
        for(int k=1; k<=n; k++){
            for(int p=1; p<=n; p++){
                if(matchData[k][i] && matchData[i][p]){
                    matchData[k][p] = true;
                }
            }
        }
    }
    for(int i=1; i<=n; i++){
        int dataCnt = 0;    
        for(int k=1; k<=n; k++){
            if(matchData[i][k]||matchData[k][i]){
                dataCnt++;
            }
        }
        if(dataCnt == n-1){
            answer++;
        }
    }
    return answer;
}

File: 2022W09/fienestar/pgr_49191.js
/*
    순위
    https://programmers.co.kr/learn/courses/30/lessons/49191
*/

function solution(n, results)
{
    let answer = 0
    let has_route = new Array(n).fill(null).map(v=>new Array(n).fill(false))
    
    for(let result of results)
        has_route[result[0]-1][result[1]-1] = true

    for(let k in has_route)
        for(let a in has_route[k])
            if(has_route[a][k])
                for(let b in has_route[k])
                    if(has_route[k][b])
                        has_route[a][b] = true
    
    for(let i in has_route){
        let win = has_route[i].filter(v=>v).length
        let lose = has_route.filter(v=>v[i]).length
        answer += (win + lose + 1 == n)
    }
    return answer
}